File: Rookie2EngineerBatch8-BE/schemas/asset.py
from pydantic import BaseModel, root_validator, field_validator
from datetime import date
from enums.shared.location import Location
from enums.asset.state import AssetState
from typing import Optional
from schemas.category import CategoryRead

class AssetHistory(BaseModel):
    """Asset history model"""
    id: int
    assign_date: date
    assigned_to: str
    assigned_by: str
    return_date: Optional[date] = None

class AssetBase(BaseModel):
    asset_name: str
    

class AssetCreate(AssetBase):
    category_id: int
    specification: str
    installed_date: date
    asset_state: AssetState

class AssetRead(AssetBase):
    id: int
    asset_code: str
    specification: str
    installed_date: date
    asset_state: AssetState
    asset_location: Location
    category: CategoryRead

class AssetUpdate(BaseModel):
    asset_name: Optional[str] = None
    specification: Optional[str] = None
    installed_date: Optional[date] = None
    asset_state: Optional[AssetState] = None

    @root_validator(pre=True)
    def check_at_least_one_field(cls, values):
        if not any(values.get(field) is not None for field in ['asset_name', 'specification', 'installed_date', 'asset_state']):
            raise ValueError('At least one field must be provided fr update')
        return values
    
    @field_validator('installed_date')
    def check_installed_date(cls, v):
        if v > date.today():
            raise ValueError('Installed date cannot be in the future')
        return v

File: Rookie2EngineerBatch8-FE/src/types/asset.ts
import { SortDirection, AssetSortOption, AssetState } from "./enums";
import { CategoryDetail } from "./category";

export interface AssetBase {
    id: number;
    asset_code: string;
    asset_name: string;
    category: CategoryDetail;
    asset_state: AssetState;
}

export interface AssetDetail extends AssetBase{
    specification: string;
    installed_date: Date;
    asset_location: string;
}

export interface AssetHistory {
    id: number,
    assign_date: Date;
    assigned_to: string;
    assigned_by: string;
    return_date: Date | null;
}

export interface AssetListQueryParams {
  page: number;
  state: AssetState[] | null;
  category: string | null;
  search: string | null;
  sortBy: AssetSortOption | null;
  sortDirection: SortDirection | null;
}

export interface AssetUpdate extends Pick<AssetBase, "asset_name" | "asset_state"> {
    specification: string;
    installed_date: Date;
}

export const AssetUpdatableStates = Object.values(AssetState).filter((state)=> state != AssetState.ASSIGNED) 
export interface Asset {
    asset_name?: string;
    category_id?: number;
    specification?: string;
    installed_date?: Date;
    state?: StateType; 
}

export interface AssetCreate {
    asset_name?: string;
    category_id?: number;
    specification?: string;
    installed_date?: Date;
    asset_state?: StateType;
}

export const StateType = {
    AVAILABLE: "Available",
    NOT_AVAILABLE: "Not Available",
} as const;

export type StateType = typeof StateType[keyof typeof StateType];

export interface AssetDetail {
    id: number;
    asset_code: string;
    asset_name: string;
    category: CategoryDetail;
    specification: string;
    installed_date: Date;
    state?: AssetState;
    location?: string;
}

export interface AssetUpdateForm {
    asset_code: string;
    asset_name: string;
    category: string;
    specification?: string;
    installed_date?: string;
    state?: string;
    location?: string;
}
